components/prize_pool/commons/prize_pool.lua

```diff
@@ -0,0 +1,352 @@
+	for placementIndex = 1, math.huge do
+		local placementInput = Json.parseIfString(args[placementIndex])
+		if not placementInput then
+			break
+		end
+
+		local placement = {}
+		-- Parse place, explicit
+		if placementInput.place then
+			local places = Table.mapValues(mw.text.split(placementInput.place, '-'), tonumber)
+			placement.placeStart = places[1]
+			placement.placeEnd = places[2] or places[1]
+		end
+
+		-- Parse prizes
+		placement.prizes = self:_readPrizeRewards(placementInput)
+
+		-- Parse opponents in the placement
+		placement.opponents = {}
+		for opponentIndex = 1, math.huge do
+			local opponentInput = Json.parseIfString(placementInput[opponentIndex])
+			local opponent = {opponentData = {}, prizes = {}, data = {}}
+			if not opponentInput then
+				-- If given a range of opponents, add them all, even if they're missing from the input
+				if not placementInput.place or placement.placeStart + opponentIndex > placement.placeEnd + 1 then
+					break
+				else
+					opponent.opponentData = Opponent.tbd()
+				end
+			else
+				-- Parse Opponent Data
+				opponentInput.type = opponentInput.type or self.opponentType
+				opponent.opponentData = self:_parseOpponentArgs(opponentInput, opponentInput.date)
+
+				-- Parse special prizes for this opponent
+				opponent.prizes = self:_readPrizeRewards(opponentInput)
+
+				-- Parse additional data (groupscore, last opponent etc)
+				opponent.data = self:_readSpecialData(opponentInput)
+
+				-- Set the date
+				opponent.date = opponentInput.date
+			end
+			table.insert(placement.opponents, opponent)
+		end
+
+		-- Parse place, implicit
+		if not placementInput.place then
+			placement.placeStart = currentPlace + 1
+			placement.placeEnd = currentPlace + #placement.opponents
+		end
+
+		assert(placement.placeStart and placement.placeEnd, 'readPlacements: Invalid |place= provided.')
+		assert(#placement.opponents > placement.placeEnd - placement.placeStart, 'readPlacements: Too many opponents')
+		currentPlace = placement.placeEnd
+		table.insert(placements, placement)
+	end
+
+	return placements
+end
+
+function PrizePool:_parseOpponentArgs(input, date)
+	local opponentArgs = Json.parseIfTable(input) or (type(input) == 'table' and input or {input})
+	opponentArgs.type = opponentArgs.type or self.opponentType
+	opponentArgs.date = date or self.date
+	assert(Opponent.isType(opponentArgs.type), 'Invalid type')
+	local opponentData = Opponent.readOpponentArgs(opponentArgs) or Opponent.tbd()
+	return Opponent.resolve(opponentData, date)
+end
+
+function PrizePool:setOption(option, value, list)
+	list = list or self.options
+	list[option] = value
+	return self
+end
+
+-- TODO extract parts of this to the future Enum class
+local CUSTOM_TYPES_OFFSET = 100
+local CUSTOM_TYPES_USED = 0
+function PrizePool:addCustomPrizeType(enum, data)
+	prizeTypes[enum] = CUSTOM_TYPES_OFFSET + CUSTOM_TYPES_USED
+	CUSTOM_TYPES_USED = CUSTOM_TYPES_USED + 1
+	prizeData[prizeTypes[enum]] = data
+	return self
```

Comment:
This is hard to follow, add doc

Reply:
Reworked entirely as the enum has been dropped.